x-pack/plugins/security_solution/public/common/utils/global_query_string/index.ts

```diff
@@ -0,0 +1,133 @@
+import { parse, ParsedQuery, stringify } from 'query-string';
+import { useEffect, useMemo } from 'react';
+
+import { url } from '@kbn/kibana-utils-plugin/public';
+import { isEmpty, pickBy } from 'lodash/fp';
+import { useHistory } from 'react-router-dom';
+import {
+  decodeRisonUrlState,
+  encodeRisonUrlState,
+  getParamFromQueryString,
+  getQueryStringFromLocation,
+} from '../../components/url_state/helpers';
+import { useShallowEqualSelector } from '../../hooks/use_selector';
+import { getStore } from '../../store';
+import { globalUrlParamActions, globalUrlParamSelectors } from '../../store/global_url_param';
+import { useRouteSpy } from '../route/use_route_spy';
+import { getLinkInfo } from '../../links';
+import { SecurityPageName } from '../../../app/types';
+
+interface RegisterUrlParams<State> {
+  urlParamKey: string;
+}
+
+/**
+ * Adds UrlParamKey and the initial value to redux store.
+ */
+export const registerUrlParam = <State>({
+  urlParamKey,
+}: RegisterUrlParams<State>): {
+  state: State | null;
+  deregister: () => void;
+} => {
+  const initialValue = getParamFromQueryString(
+    getQueryStringFromLocation(window.location.search),
+    urlParamKey
+  );
+
+  const store = getStore();
+  store?.dispatch(
+    globalUrlParamActions.registerUrlParam({ key: urlParamKey, initialValue: initialValue ?? null })
+  );
+
+  const deregister = () => {
+    store?.dispatch(globalUrlParamActions.deregisterUrlParam({ key: urlParamKey }));
+  };
+
+  return { state: decodeRisonUrlState<State>(initialValue ?? undefined), deregister };
+};
+
+interface UpdateUrlParams<State> {
+  urlParamKey: string;
+  value: State | null;
+}
+
+/**
+ * Updates URL parameters in the url.
+ *
+ * Make sure to call `registerUrlParam` before calling this function.
+ */
+export const updateUrlParam = <State>({ urlParamKey, value }: UpdateUrlParams<State>) => {
+  const store = getStore();
+  const encodedValue = value !== null ? encodeRisonUrlState(value) : null;
+
+  store?.dispatch(globalUrlParamActions.updateUrlParam({ key: urlParamKey, value: encodedValue }));
+};
+
+/**
+ * Generates the global query string for security solutions links.
+ */
+export const useGlobalQueryString = (): string => {
+  const globalUrlParam = useShallowEqualSelector(globalUrlParamSelectors.selectGlobalUrlParam);
+
+  const globalQueryString = useMemo(
+    () => encodeQuerySrting(pickBy((value) => !isEmpty(value), globalUrlParam)),
+    [globalUrlParam]
+  );
+
+  return globalQueryString;
+};
+
+/**
+ * - It hides / shows the global query depending on the page.
+ * - It updates the URL when globalUrlParam store updates.
+ */
+export const useSyncGlobalQueryString = () => {
+  const history = useHistory();
+  const [{ pageName }] = useRouteSpy();
+  const globalUrlParam = useShallowEqualSelector(globalUrlParamSelectors.selectGlobalUrlParam);
+
+  useEffect(() => {
+    const linkInfo = getLinkInfo(pageName as SecurityPageName) ?? { skipUrlState: true };
```

Comment:
instead of casting, can we change the `RouteSpyState.pageName` type to `SecurityPageName`?

Reply:
Sure. I checked, and most changes required to comply with the new type are on test files. So I am adding it to the PR.